File: src/wbjp/strategy/samples/rsi_reversion.py
"""RSI 逆張り戦略。

RSI が売られすぎ圏に沈んだら買い、買われすぎ圏に浮いたら売り。

性格: レンジ相場で機能し、強いトレンドでは逆行し続けて損を膨らませる。
そこで ADX でトレンドの強さを測り、**強トレンド時は発言を控える**
フィルタを入れてある。順張り戦略との組み合わせを前提とした設計。
"""

from __future__ import annotations

from typing import ClassVar

import polars as pl

from wbcore.domain.models import Signal
from wbcore.indicators.ohlcv import adx, rsi
from wbjp.strategy.base import IndicatorStrategy, StrategyContext


class RsiReversionStrategy(IndicatorStrategy):
    name: ClassVar[str] = "rsi_reversion"

    def __init__(
        self,
        period: int = 14,
        oversold: float = 30.0,
        overbought: float = 70.0,
        adx_period: int = 14,
        max_adx: float = 40.0,
    ) -> None:
        if not 0 < oversold < overbought < 100:
            raise ValueError(
                f"0 < oversold < overbought < 100 を満たすこと: {oversold}, {overbought}"
            )
        self.period = period
        self.oversold = oversold
        self.overbought = overbought
        self.adx_period = adx_period
        self.max_adx = max_adx
        # ADX は内部で2段の Wilder 平滑化を行うため、その分の助走が要る
        self.warmup_bars = max(period, adx_period * 3) + 1

    def indicators(self) -> list[pl.Expr]:
        return [rsi(self.period), *adx(self.adx_period)]

    def evaluate(self, symbol: str, frame: pl.DataFrame, ctx: StrategyContext) -> Signal | None:
        latest = frame.row(-1, named=True)
        rsi_value = latest[f"rsi_{self.period}"]
        adx_value = latest[f"adx_{self.adx_period}"]

        if rsi_value is None:
            return None

        # 強いトレンドが出ている間は逆張りしない。
        # 逆張り戦略が最も損をするのがこの局面なので、黙る判断を明示する。
        if adx_value is not None and adx_value > self.max_adx:
            return None

        if rsi_value <= self.oversold:
            depth = (self.oversold - rsi_value) / self.oversold
            return Signal(
                self.name,
                symbol,
                direction=1.0,
                confidence=min(1.0, 0.4 + depth),
                reason=f"RSI {rsi_value:.1f} が売られすぎ圏（≦{self.oversold:g}）",
                meta={"rsi": rsi_value, "adx": adx_value},
            )

        if rsi_value >= self.overbought:
            depth = (rsi_value - self.overbought) / (100 - self.overbought)
            return Signal(
                self.name,
                symbol,
                direction=-1.0,
                confidence=min(1.0, 0.4 + depth),
                reason=f"RSI {rsi_value:.1f} が買われすぎ圏（≧{self.overbought:g}）",
                meta={"rsi": rsi_value, "adx": adx_value},
            )

        return None

    def describe(self) -> str:
        return (
            f"{self.name}(period={self.period}, "
            f"oversold={self.oversold:g}, overbought={self.overbought:g})"
        )

File: pkg/wbjp/strategy/rsi_reversion.go
package strategy

import (
	"fmt"
	"math"

	"github.com/lovemoneyhotspring/jstock-go/pkg/wbcore/domain"
)

// RSIReversion は RSI 逆張り戦略。
//
// RSI が売られすぎ圏に沈んだら買い、買われすぎ圏に浮いたら売り。レンジ相場で
// 機能し、強いトレンドでは逆行し続けて損を膨らませる。そこで ADX でトレンドの
// 強さを測り、強トレンド時は発言を控える。順張り戦略との組み合わせが前提。
type RSIReversion struct {
	Period     int
	Oversold   float64
	Overbought float64
	ADXPeriod  int
	MaxADX     float64

	warmup int
}

// NewRSIReversion は RSI 逆張り戦略を作る。
func NewRSIReversion(period int, oversold, overbought float64, adxPeriod int, maxADX float64) (*RSIReversion, error) {
	if period <= 0 {
		period = 14
	}
	if adxPeriod <= 0 {
		adxPeriod = 14
	}
	if !(0 < oversold && oversold < overbought && overbought < 100) {
		return nil, fmt.Errorf("0 < oversold < overbought < 100 を満たすこと: %g, %g", oversold, overbought)
	}
	return &RSIReversion{
		Period:     period,
		Oversold:   oversold,
		Overbought: overbought,
		ADXPeriod:  adxPeriod,
		MaxADX:     maxADX,
		// ADX は内部で2段の Wilder 平滑化を行うため、その分の助走が要る
		warmup: maxInt(period, adxPeriod*3) + 1,
	}, nil
}

func (r *RSIReversion) Name() string    { return "rsi_reversion" }
func (r *RSIReversion) WarmupBars() int { return r.warmup }
func (r *RSIReversion) Describe() string {
	return fmt.Sprintf("%s(period=%d, oversold=%g, overbought=%g)", r.Name(), r.Period, r.Oversold, r.Overbought)
}

func (r *RSIReversion) OnBars(ctx *Context) ([]domain.Signal, error) {
	return eachSymbol(ctx, r.warmup, r.evaluate), nil
}

func (r *RSIReversion) evaluate(symbol string, v View) *domain.Signal {
	rsiValue := last(v.RSI(r.Period))
	adxValue := last(v.ADX(r.ADXPeriod))
	if math.IsNaN(rsiValue) {
		return nil
	}

	// 強いトレンドが出ている間は逆張りしない。
	// 逆張り戦略が最も損をするのがこの局面なので、黙る判断を明示する。
	if !math.IsNaN(adxValue) && adxValue > r.MaxADX {
		return nil
	}

	meta := map[string]any{"rsi": rsiValue, "adx": adxValue}

	if rsiValue <= r.Oversold {
		depth := (r.Oversold - rsiValue) / r.Oversold
		return signal(r.Name(), symbol, 1.0, math.Min(1.0, 0.4+depth),
			fmt.Sprintf("RSI %.1f が売られすぎ圏（≦%g）", rsiValue, r.Oversold), meta)
	}
	if rsiValue >= r.Overbought {
		depth := (rsiValue - r.Overbought) / (100 - r.Overbought)
		return signal(r.Name(), symbol, -1.0, math.Min(1.0, 0.4+depth),
			fmt.Sprintf("RSI %.1f が買われすぎ圏（≧%g）", rsiValue, r.Overbought), meta)
	}
	return nil
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}
